src/Language/Haskell/TH/Syntax/Compat.hs

```diff
@@ -341,3 +388,162 @@ instance Quote m => Quasi (QuoteToQuasi m) where
 #if MIN_VERSION_template_haskell(2,16,0)
   qReifyType          = qtqError "qReifyType"
 #endif
+
+-------------------------------------------------------------------------------
+-- Code
+-------------------------------------------------------------------------------
+
+-- $code
+-- The 'Code' type (first proposed in
+-- <https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0195-code-texp.rst GHC Proposal 195>)
+-- was introduced in @template-haskell-2.17.0.0@. This module defines a version
+-- of 'Code' that is backward-compatible with older @template-haskell@
+-- releases and is forward-compatible with the existing 'Code' class.
+-- In addition to 'Code', this module also backports the functions in
+-- "Language.Haskell.TH.Syntax" that manipulate 'Code' values.
+--
+-- One troublesome aspect of writing backwards-compatible code involving 'Code'
+-- is that GHC 9.0 changed the types of typed Template Haskell splices. Before,
+-- they were of type @'Q' ('TExp' a)@, but they are now of type @'Code' 'Q' a@.
+-- The 'IsCode' class can be used to paper over the difference between these
+-- two types. For more details, consult the Haddocks for 'IsCode'.
+
+#if MIN_VERSION_template_haskell(2,9,0)
+-- | A class that allows one to smooth over the differences between
+-- @'Code' 'm' a@ (the type of typed Template Haskell quotations on
+-- @template-haskell-2.17.0.0@ or later) and @'m' ('TExp' a)@ (the type of
+-- typed Template Haskell quotations on older versions of @template-haskell@).
+-- Here are two examples that demonstrate how to use each method of 'IsCode':
+--
+-- @
+-- &#123;-&#35; LANGUAGE TemplateHaskell &#35;-&#125;
+--
+-- import "Language.Haskell.TH"
+-- import "Language.Haskell.TH.Syntax.Compat"
+--
+-- -- 'toCode' will ensure that the end result is a 'Code', regardless of
+-- -- whether the quote itself returns a 'Code' or a 'TExp'.
+-- myCode :: 'Code' 'Q' Int
+-- myCode = 'toCode' [|| 42 ||]
+--
+-- -- 'fromCode' will ensure that the input 'Code' is suitable for splicing
+-- (i.e., it will return a 'Code' or a 'TExp' depending on the
+-- @template-haskell@ version in use).
+-- fortyTwo :: Int
+-- fortyTwo = $$('fromCode' myCode)
+-- @
+class IsCode q a c | c -> a q where
+  -- | Convert something to a 'Code'.
+  toCode   :: c -> Code q a
+  -- | Convert to something from a 'Code'.
+  fromCode :: Code q a -> c
+
+instance Quote q => IsCode q a (Code q a) where
```

Comment:
Another comment is due how `Quote` is defined here, you cannot simply write

```haskell
instance (TH.Lift a, Quote q, Num a) => Num (Code q a) where
    fromInteger x = liftTyped (fromInteger x)

    x + y = toCode [|| $$(fromCode x) + $$(fromCode y) ||]
    x - y = toCode [|| $$(fromCode x) - $$(fromCode y) ||]
    x * y = toCode [|| $$(fromCode x) * $$(fromCode y) ||]

    abs    x = toCode [|| abs $$(fromCode x) ||]
    negate x = toCode [|| negate $$(fromCode x) ||]
    signum x = toCode [|| signum $$(fromCode x) ||]
```

In fact, I don't see how this is possible to do on older GHCs at all. `unsafeQToQuery` doesn't (and cannot) work here.

Reply:
I'm not sure which issue in particular you are referring to with this comment. I can see at least two possible issues:

1. The fact that `fromInteger x = liftTyped (fromInteger x)` doesn't typecheck on pre-2.17 versions of `template-haskell`. This is easily fixable: just use `liftTypedQuote` instead, since that backports the modern type of `liftTyped` back to old versions of `template-haskell`.
2. The fact that something like `x + y = toCode [|| $$(fromCode x) + $$(fromCode y) ||]` doesn't typecheck. Indeed, I think the culprit is this instance:

   ```hs
   instance texp ~ Syntax.TExp a => IsCode Q a (Q texp) where
     toCode   = liftCode
     fromCode = examineCode
   ```
   
   This hard-codes the use of `Q`, which precludes the use of an arbitrary `Quote` instance `q`. It's tempting to try changing the instance to this:
   
   ```hs
   instance Quote q => IsCode q a (q (TExp a)) where
     toCode   = liftCode
     fromCode = examineCode
   ```
   
   But that won't satisfy the functional dependencies for `IsCode`. Any ideas?

Reply:
changing instance won't help. `$$( ... )` wants `Q (TExp a)`. Try inlining `examineCode` for `fromCode` in the RHS of `x + y`

Reply:
Ah, right. Of course that wouldn't work for old GHCs, since `Quote` didn't even exist back then (just `Q`).

Unfortunately, I don't have any ideas for making this work. But then again, `Quote` is a new feature in `template-haskell-2.17.0.0`, so we may just have to live with not being able to write this sort of code in a way that's backwards-compatible. At the very least, I could document this as a shortcoming of this library.

Reply:
It's possible to write code like that by using different "compat" `type Quote m = m ~ Q`.

For my use case it seems right approach. I don't claim it's "the" right approach.

Reply:
Hm, that's an interesting approach. May I ask why you wouldn't just program directly against `Q` at that point?

Reply:
I can provide more general types for combinators with GHC-9.0

Yet, this got me thinking whether I can use your compat `Quote`, with `unsafeTExpCoerceQuote` and constructing stuff manually using `AppE` etc. constructors. I'll try.

Reply:
OK. If need be, we can always introduce something like this for your use case (better naming suggestions welcome):

```hs
#if MIN_VERSION_template_haskell(2,17,0)
class Quote q => QuoteCompat q
instance Quote q => QuoteCompat q
#else
class (q ~ Q) => QuoteCompat q
instance (q ~ Q) => QuoteCompat q
#endif
```

Like `SpliceQ`, this would need to be used with care, since it uses a conditional API. But it wouldn't create much additional maintenance burden, at least.

Reply:
I'll try that today too.

Is there a reason why it's not `type QuoteCompat = Quote` with newer `template-haskell`. Is there some subtle difference when there are no type-arguments, or is it just for uniformity?

Reply:
So I can use this `Quote`

```haskell
instance (TH.Lift a, Compat.Quote q, Num a) => Num (Code q a) where
    fromInteger x = liftTypedQuote (fromInteger x)

#if MIN_VERSION_template_haskell(2,17,0)
    x + y = [|| $$x + $$y ||]
    x - y = [|| $$x - $$y ||]
    x * y = [|| $$x * $$y ||]

    abs    x = [|| abs $$x ||]
    negate x = [|| negate $$x ||]
    signum x = [|| signum $$x ||]
#else
    x + y = unsafeCodeCoerce $ plus <$> unTypeCode x <*> unTypeCode y where
        plus x' y' = TH.VarE '(+) `TH.AppE` x' `TH.AppE` y'

    -- rest methods same
#endif
```

And as nothing tricky happens in splices, they work:

```haskell
*Staged.Commons> $$(fromCode ((1 :: Code Q Int) + 2)) :: Int
<interactive>:10:4-35: Splicing expression
    fromCode ((1 :: Code Q Int) + 2) ======> ((+) 1) 2
3
```

Reply:
Random thought. One could want to have something stronger than `Lift`, to perform extra checks when doing `fromInteger :: Integer -> Code q Natural`, or similarly for `fromString :: String -> Code q ByteString`.

That **isn't** a problem, as `Quote` is too weak to do such extra stuff. So for both 9.0 and older GHCs one have to choose between (real or fake `Quote` and no checks) or `Quasi and checks`.

That to justify, that even I do some `unsafe*` stuff, we assume that values are either `lift`ed or made with `[[ ... ||`, thus don't have extra `Q` effects. Or if they have, then I don't do `unsafeQToQuote` (in `+` definition).

I think pieces are slowly starting to fit together.

Reply:
> Is there a reason why it's not type QuoteCompat = Quote with newer template-haskell. Is there some subtle difference when there are no type-arguments, or is it just for uniformity?

I did this for slightly more uniformity, since defining `QuoteCompat` as a class would let you partially apply it on all versions of `template-haskell` (whereas `type QuoteCompat q = (q ~ Q)` wouldn't be partially applicable). But now that I think about it, we'll need to use a type synonym on at least one version of GHC, since GHC 7.0 refuses to compile `class (q ~ Q) => QuoteCompat`:

```
Bug.hs:6:1:
    Alas, GHC 7.0 still cannot handle equality superclasses: q ~ Q
    In the context: (q ~ Q)
    While checking the super-classes of class `QuoteCompat'
    In the class declaration for `QuoteCompat'
```

Bummer. And to really add insult to injury, this won't parse on GHC 7.0 or 7.2:

```hs
type QuoteCompat = (~) Q
```

> So I can use this `Quote`
> 
> ```haskell
> instance (TH.Lift a, Compat.Quote q, Num a) => Num (Code q a) where
>     fromInteger x = liftTypedQuote (fromInteger x)
> 
> #if MIN_VERSION_template_haskell(2,17,0)
>     x + y = [|| $$x + $$y ||]
>     x - y = [|| $$x - $$y ||]
>     x * y = [|| $$x * $$y ||]
> 
>     abs    x = [|| abs $$x ||]
>     negate x = [|| negate $$x ||]
>     signum x = [|| signum $$x ||]
> #else
>     x + y = unsafeCodeCoerce $ plus <$> unTypeCode x <*> unTypeCode y where
>         plus x' y' = TH.VarE '(+) `TH.AppE` x' `TH.AppE` y'
> 
>     -- rest methods same
> #endif
> ```

That's quite brutal. :)

> I think pieces are slowly starting to fit together.

I had a bit of a hard time following https://github.com/haskell-compat/th-quote/pull/1#discussion_r463958205, so would you mind summarizing what recommendations you would make for making `th-compat` suitable for your needs?

Reply:
We can get further if this PR is divided into pieces. Something which is good are

- Code and it's functions.

- Let us have, **this is the only change I think we should do to this PR now**

```haskell
liftTypedQuote :: (TH.Lift a, Compat.Quote q) => a -> Code q a
#if MIN_VERSION_template_haskell(2,17,0)
liftTypedQuote = TH.liftTyped
#elif MIN_VERSION_template_haskell(2,16,0)
liftTypedQuote = Compat.liftCode . Compat.unsafeQToQuote  . Compat.lifTyped
#else
liftTypedQuote = Compat.unsafeCodeCoerce . Compat.liftQuote
#endif
```

---

About the rest, I'm not sure anymore.

I think we should leave `QuoteCompat` and `SpliceQ` out for now.

---

I want to give it a try and rewrite whole `staged-commons` in

```haskell
stagedId :: Quote q => Code q a -> Code a q
#if MIN_VERSION_template_haskell(2,17,0)
stagedId x = [|| id $$x ||]
#else
stagedId x = unsafeCodeCoerce $ mk <$> unTypeCode x where
  mk x' = VarE 'id `AppE` x'
#endif
```

style.

If that works, then I won't need `QuoteCompat` nor `SpliceQ` there.
Which would be good thing, even the implementation of
`staged-commons` become complete horror.

Reply:
OK. At the very least, I'll make a note of the limitations surrounding the backported `Quote` and typed TH, and then we can revisit the issue later if need be.

Reply:
I've added a "Limitations of `IsCode`" section to the Haddocks.